### DISCUSSION TITLE
Add navigation to pytest fixtures

### DISCUSSION THREAD

**Discussion by TylerYep** [microsoft/pylance-release#2576] [Enhancement]
When using pytest, tests can use fixtures that allow mocked objects or functions to be easily reused in tests. They look like parameters, but through some Pytest magic, they are invoked and passed into the test function.

My feature request is that Cmd+clicking on the parameter of a test function takes me to file or definition where the fixture is defined.

```
def test_with_fixture(fixture_1: dict[str, str]) -> None:
    ...
```
Cmd+clicking on the `fixture_1` parameter should take me to:
```
@pytest.fixture
def fixture_1() -> dict[str, str]:
    return {"data": "stuff"}
```

Reference: https://docs.pytest.org/en/6.2.x/fixture.html

**Comment by macintacos** [comment]
Will just say that on even medium-sized test suites, with fixtures spread out across various files, this can become a real pain. Would be fantastic to see this implemented!

**Comment by ItamarShDev** [comment]
Some existing extensions for reference: 
- https://marketplace.visualstudio.com/items?itemName=Cameron.vscode-pytest
- https://marketplace.visualstudio.com/items?itemName=nickmillerdev.pytest-fixtures

**Comment by jwilges** [comment]
Having IntelliSense support for this would be nice if navigation is also on the table. I started a discussion about getting completion for pytest fixtures in `vscode-python` here: https://github.com/microsoft/vscode-python/discussions/20471#discussion-4724128.

But, I think that was the wrong place to launch this discussion. @brettcannon referred me here, which makes sense.

If the Pylance extension is not open source, is there at least a recommended mechanism outside folks could tinker with to start hacking together a reference implementation for fixture support aside from creating a standalone extension?

Or would providing IntelliSense + code navigation via an extension be useful? The extensions linked above can serve as a reference but they are not what I would think would be "production grade" in the sense that they don't consider/support workspace features like `python.envFile` which may in some cases be required to initialize the correct environment for pytest to run (if using `pytest --fixtures` to scan for fixture names/paths). For out of the box support on this, I'd hope the implementation would take into account all meaningful hints from standard VS Code workspace/user settings e.g. `pytest` path/arguments, environment variables, virtual environment particulars, and other odds and ends I'm surely not thinking about off hand.

**Reply by rchiodo** [reply]
We're actually investigating this idea internally at the moment.
https://github.com/microsoft/pylance-release/issues/3727

The prototype we built was based on pyright, so theoretically you could write an extension that did that too.

**Reply by jwilges** [reply]
This is awesome to hear @rchiodo, I'll try to take a closer look at that thread to see what I can glean. Given `pytest`'s dominance in Python testing, any improvements in this area will be well received by many, I'm sure. 

> The prototype we built was based on pyright, so theoretically you could write an extension that did that too.

I'm definitely interested in exploring some form of improvement in this domain (fixture completion / navigation) if there's a way outsiders can help. I wouldn't want to duplicate internal efforts though if there are better ways to help. Appreciate the replies from you all so far!

**Comment by danhje** [comment]
On the landing page of one of the extensions linked in a previous comment, it says that "As of Visual Studio Code version [1.76](https://code.visualstudio.com/updates/v1_76#_improved-intellisense-support-for-pytest), Pylance natively supports pytest fixture detection", but it doesn't seem to work for me. "Go to definition" just highlights the parameter, not the fixture. Anyone else?

**Reply by rchiodo** [reply]
Do you have this setting enabled?

![image](https://github.com/user-attachments/assets/c68aadc3-82ee-4313-9b00-12890162041a)

There is one bug at the moment where it doesn't work for test classes though:
https://github.com/microsoft/pylance-release/issues/6292
